### ISSUE TITLE
URL equality/passing None to URL.build

### ISSUE THREAD

**Issue by PromyLOPh** [aio-libs/yarl#275]
The equality operator (==, \_\_eq\_\_) does not work properly when passing `None` to URL.build. Example:

```python
a, b = URL.build(scheme='http', host='0'), URL.build(scheme='http', host='0', fragment=None)
assert a == b # this is False
assert repr (a) == repr (b) # this is True, because "URL('http://0')" == "URL('http://0')"
```

Passing an empty string works as expected. This is very confusing, because repr produces exactly the same representation.

**Comment by asvetlov** [comment]
That's interesting.
`URL.build` declares `fragment` as non-optional `str` parameter with default to empty string.
Why do you pass `None`? `None` for the argument is a grey zone: sometimes it works but there is no any commitment for this value type.

**Comment by webknjaz** [comment]
I think `None` should correspond to missing fragment with no hash `#` symbol while empty string would be trailing hash in URL. `http://some.com/url` vs. `http://some.com/url#`

**Comment by asvetlov** [comment]
`urllib.parse.urlparse` doesn't distinguish `http://some.com/url` and `http://some.com/url#`.
It returns an empty string fragment for both cases.
We can return to the question if will implement a custom HTTP parser but now there is nothing to do.

**Comment by PromyLOPh** [comment]
I’m using hypothesis to generate test data. More specifically a dict which is then passed to yarl using URL.build(**dict). And I was simply assuming None would be mapped to the default value, which is apparently wrong (judging from https://yarl.readthedocs.io/en/latest/api.html#yarl.URL.fragment). This
is fine for me, however the behavior of == and repr() in this case are very confusing and thus None should either be rejected explicitly (TypeError) or silently converted to '' to prevent this situation.

**Comment by asvetlov** [comment]
TypeError sounds good to me.
Would you prepare a pull request?

**Comment by asvetlov** [comment]
Fixed by #276
